### PR TITLE
Rulefit - opt3

### PR DESCRIPTION
**Pull request by koniecsveta**

### CODE REVIEW

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java [hunk 1]
```diff
@@ -75,23 +70,48 @@ else if (filteredRule.size() > 1) {
     }
 }
 
+class Transform extends MRTask<Transform> {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java:73]: make inner class

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java [hunk 2]
```diff
@@ -75,23 +70,48 @@ else if (filteredRule.size() > 1) {
...
+    String[] _names;
+
+    Transform(Rule[] rules, String[] names) {
+        _rules = rules;
+        _names = names;
+    }
+
+    @Override
+    public void map(Chunk[] cs, NewChunk[] ncs) {
+        byte[][] bytes = new byte[_rules.length][];
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java:84]: you actually only need one bytearray, should be argument sent to rule.transform()

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java [hunk 3]
```diff
@@ -75,23 +70,48 @@ else if (filteredRule.size() > 1) {
...
+            // write result to NewChunk
+            for (int j = 0; j < bytes[i].length; j++)
+                ncs[i].addNum(bytes[i][j]);
+        }
+    }
+}
+
+
+
 class Decoder extends MRTask<hex.rulefit.Decoder> {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java:98]: would also move to inner class

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java [hunk 4]
```diff
@@ -75,23 +70,48 @@ else if (filteredRule.size() > 1) {
...
+    Transform(Rule[] rules, String[] names) {
+        _rules = rules;
+        _names = names;
+    }
+
+    @Override
+    public void map(Chunk[] cs, NewChunk[] ncs) {
+        byte[][] bytes = new byte[_rules.length][];
+        for (int i = 0; i < _rules.length; i++) {
+            bytes[i] = _rules[i].transform(cs);
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java:86]: would add Arrays.fill(out, (byte) 1);

File: h2o-algos/src/main/java/hex/rulefit/Rule.java [hunk 5]
```diff
@@ -150,6 +159,19 @@ public CategoricalThreshold(String[] catThreshold, int[] catThresholdNum) {
         return Math.abs(coefficient);
     }
 
+    byte[] ruleReducer(byte[][] bytes) {
+        byte[] ncs = MemoryManager.malloc1(bytes[0].length);
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/Rule.java:163]: move to RuleEnsamble

File: h2o-algos/src/main/java/hex/rulefit/Rule.java [hunk 6]
```diff
@@ -150,6 +159,19 @@ public CategoricalThreshold(String[] catThreshold, int[] catThresholdNum) {
...
+        byte newVal;
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/Rule.java:164]: can be moved to inside forloop

File: h2o-algos/src/main/java/hex/rulefit/Condition.java [hunk 7]
```diff
@@ -72,10 +122,23 @@ public int hashCode() {
...
+            Chunk col = cs[Condition.this.featureIndex];
+            byte[] out = MemoryManager.malloc1(col.len());
+            Arrays.fill(out, (byte) 1);
+            map(cs, out);
+            for (byte b : out) {
+                nc.addNum(b);
+            }
+        }
+
+        public void map(Chunk[] cs, byte[] out) {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/rulefit/Condition.java:136]: code duplicated above

File: h2o-algos/src/main/java/hex/rulefit/Rule.java (hunk 5 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Rule.java:163]: would leave this here: when you have rule composed out of three conditions, you get three binary columns by evaluating these conditions on data and rule reducer reduces them to 1 column which represents this rule

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java (hunk 3 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java (hunk 4 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/RuleEnsemble.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/Rule.java (hunk 5 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Rule.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/Rule.java (hunk 6 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Rule.java]: fixed

File: h2o-algos/src/main/java/hex/rulefit/Condition.java (hunk 7 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/rulefit/Condition.java:96]: fixed

**Comment by koniecsveta** [comment]
this is slower against https://github.com/h2oai/h2o-3/pull/5010/files

**Comment by koniecsveta** [comment]
closing this one as slow

**Pull request closed without merging**